contracts/liquidityMining/KyberFairLaunchV2.sol

```diff
@@ -0,0 +1,603 @@
+/// Allow extend or renew a pool to continue/restart the LM program
+/// When harvesting, rewards will be transferred to a RewardLocker
+/// Support multiple reward tokens, reward tokens must be distinct and immutable
+contract KyberFairLaunchV2 is IKyberFairLaunchV2, PermissionAdmin, ReentrancyGuard {
+  using SafeMath for uint256;
+  using SafeERC20 for IERC20Ext;
+
+  uint256 internal constant PRECISION = 1e12;
+
+  struct UserRewardData {
+    uint256 unclaimedReward;
+    uint256 lastRewardPerShare;
+  }
+  // Info of each user.
+  struct UserInfo {
+    uint256 amount; // How many Staking tokens the user has provided.
+    mapping (uint256 => UserRewardData) userRewardData;
+    //
+    // Basically, any point in time, the amount of reward token
+    // entitled to a user but is pending to be distributed is:
+    //
+    //   pending reward = user.unclaimAmount + (user.amount * (pool.accRewardPerShare - user.lastRewardPerShare)
+    //
+    // Whenever a user deposits or withdraws Staking tokens to a pool. Here's what happens:
+    //   1. The pool's `accRewardPerShare` (and `lastRewardTime`) gets updated.
+    //   2. User receives the pending reward sent to his/her address.
+    //   3. User's `lastRewardPerShare` gets updated.
+    //   4. User's `amount` gets updated.
+  }
+
+  struct PoolRewardData {
+    uint256 rewardPerSecond;
+    uint256 accRewardPerShare;
+  }
+  // Info of each pool
+  // poolRewardData: reward data for each reward token
+  //      rewardPerSecond: amount of reward token per second
+  //      accRewardPerShare: accumulated reward per share of token
+  // totalStake: total amount of stakeToken has been staked
+  // stakeToken: token to stake, should be the DMM-LP token
+  // generatedToken: token that has been deployed for this pool
+  // startTime: the time that the reward starts
+  // endTime: the time that the reward ends
+  // lastRewardTime: last time that rewards distribution occurs
+  struct PoolInfo {
+    uint256 totalStake;
+    address stakeToken;
+    GeneratedToken generatedToken;
+    uint32 startTime;
+    uint32 endTime;
+    uint32 lastRewardTime;
+    mapping (uint256 => PoolRewardData) poolRewardData;
+  }
+
+  // check if a pool exists for a stakeToken
+  mapping(address => bool) public poolExists;
+  // list reward tokens, use 0x0 for native token, shouldn't be too many reward tokens
+  // don't validate values or length by trusting the deployer
+  address[] public rewardTokens;
+  // contract for locking reward
+  IKyberRewardLocker public immutable rewardLocker;
+
+  // Info of each pool.
+  uint256 public override poolLength;
+  mapping(uint256 => PoolInfo) internal poolInfo;
+  // Info of each user that stakes Staking tokens.
+  mapping(uint256 => mapping(address => UserInfo)) internal userInfo;
+
+  event AddNewPool(
+    address indexed stakeToken,
+    address indexed generatedToken,
+    uint32 startTime,
+    uint32 endTime,
+    uint256[] rewardPerSeconds
```

Comment:
hmmm doesnt make sense to emit this because it'll be the keccak hash of the uint values.. same for `RenewPool` and `UpdatePool`.
Although V1 is liddat lol, hahaha

Reply:
Yeah removed from event